echo_server.py

```diff
@@ -45,15 +53,59 @@ def echo_server():
     server_socket.close()
 
 
-def response_ok(body=''):
+def resolve_uri(uri):
+    """Given a URI, return a body and content type.
+
+    If the URI is a directory, return that directory as a list as the body.
+    If the URI a file, return the contents of the file as the body
+    Content type is the file type.
+    Raises type error if uri is not a string.
+    Raises an IO error if the uri is not a file or directory.
+    Raises an IO error if cannot find a mimetype.
+    """
+    if not isinstance(uri, str):
+        raise HTTPError404('Not found')
+    uri = os.path.join(ROOT_DIR, uri)
+    extension = os.path.splitext(uri)
+    try:
+        content_type = mimetypes.types_map[extension[1]]
+    except KeyError:
+        content_type = ''
+    if os.path.isfile(uri):  # if uri is a file
+        if content_type == 'text/plain' or content_type == 'text/html':
+                with io.open(uri, 'r') as file1:
+                    body = file1.read()
+        elif content_type == 'image/jpeg' or content_type == 'image/png':
+                with io.open(uri, 'rb') as file1:
+                    body = file1.read()
+        else:
+            raise HTTPError415('Unsupported media type')
+    elif os.path.isdir(uri):  # if uri is a directory
+        content_type = 'directory'
+        dir_list = os.listdir(uri)
+        for index, item in enumerate(dir_list):
+            dir_list[index] = "<li>{}</li>".format(item)
```

Comment:
careful about mutating a list while iterating over it.  In this case it's not likely to screw you up, but it's a trap that's easy to fall into.